components/script/dom/dommatrixreadonly.rs

```diff
@@ -0,0 +1,687 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::DOMMatrixBinding::{DOMMatrixInit, DOMMatrixMethods};
+use dom::bindings::codegen::Bindings::DOMMatrixReadOnlyBinding::{DOMMatrixReadOnlyMethods, Wrap};
+use dom::bindings::codegen::Bindings::DOMPointBinding::DOMPointInit;
+use dom::bindings::error;
+use dom::bindings::error::Fallible;
+use dom::bindings::global::GlobalRef;
+use dom::bindings::js::Root;
+use dom::bindings::reflector::{reflect_dom_object, Reflectable, Reflector};
+use dom::dommatrix::DOMMatrix;
+use dom::dompoint::DOMPoint;
+use euclid::Matrix4D;
+use euclid::Point4D;
+use std::cell::Cell;
+use std::f64;
+
+#[dom_struct]
+pub struct DOMMatrixReadOnly {
+    reflector_: Reflector,
+    matrix: DOMRefCell<Matrix4D<f64>>,
+    is2D: Cell<bool>,
+}
+
+impl DOMMatrixReadOnly {
+    #[allow(unrooted_must_root)]
+    pub fn new(global: GlobalRef, is2D: bool, matrix: Matrix4D<f64>) -> Root<Self> {
+        let dommatrix = Self::new_inherited(is2D, matrix);
+        reflect_dom_object(box dommatrix, global, Wrap)
+    }
+
+    pub fn new_inherited(is2D: bool, matrix: Matrix4D<f64>) -> Self {
+        DOMMatrixReadOnly {
+            reflector_: Reflector::new(),
+            matrix: DOMRefCell::new(matrix),
+            is2D: Cell::new(is2D),
+         }
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-dommatrixreadonly
+    pub fn Constructor(global: GlobalRef) -> Fallible<Root<Self>> {
+        entries_to_matrix(&[1.0, 0.0, 0.0, 1.0, 0.0, 0.0])
+            .map(|(is2D, matrix)| {
+                Self::new(global, is2D, matrix)
+            })
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-dommatrixreadonly-numbersequence
+    pub fn Constructor_(global: GlobalRef, entries: Vec<f64>) -> Fallible<Root<Self>> {
+        entries_to_matrix(&entries[..])
+            .map(|(is2D, matrix)| {
+                Self::new(global, is2D, matrix)
+            })
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-frommatrix
+    pub fn FromMatrix(global: GlobalRef, other: &DOMMatrixInit) -> Fallible<Root<Self>> {
+        dommatrixinit_to_matrix(&other)
+            .map(|(is2D, matrix)| {
+                Self::new(global, is2D, matrix)
+            })
+    }
+
+    fn to_DOMMatrix(&self) -> Root<DOMMatrix> {
```

Comment:
@nox In your previous feedback, you suggested that I move this to `DOMMatrix` and rename it `from_readonly`. However, I have struggled to find a way to make that work. Here are my comments from the previous discussion:

=============

Are you suggesting an implementation like one of these?
```Rust
    pub fn from_readonly(ro: &DOMMatrixReadOnly) -> Root<DOMMatrix> {
        Self::new(ro.global().r(), ro.is2D.get(), ro.matrix.borrow().clone())
    }
```
I'm unsure about the global ref. I would also have to make the DOMMatrixReadOnly fields public, which isn't permitted without adding a pragma.

I tried explicitly downcasting, but I'd still have to move the value, which can't be done since it's borrowed already and DOMMatrixReadOnly cannot be cloned.
```Rust
    pub fn from_readonly(ro: &DOMMatrixReadOnly) -> Root<DOMMatrix> {
        reflect_dom_object(box ro.downcast().unwrap(), ro.global().r(), Wrap)
    }
```
I also tried skipping the root and letting the consumer deal with that, but it doesn't work because parent can't be a reference, so it's the same problem.
```Rust
    pub fn from_readonly(ro: &DOMMatrixReadOnly) -> DOMMatrix {
        DOMMatrix {
            parent: ro
        }
    }
```